thresholdkey.go

```diff
@@ -241,34 +273,66 @@ type PartialDecryption struct {
 	Decryption *big.Int
 }
 
+// A non-interactive ZKP based on the Fiat–Shamir heuristic. This algorithm
+// proves that the decryption server indeed raised secret to his secret exponent
+// (`ThresholdPrivateKey.Share`) by comparison with the public verification key
+// (`ThresholdKey.Vi`). Recall that v_i = v^(delta s_i).
+//
+// This is essentialy a protocol for the equality of discrete logs,
+// log_{c^4}(c_i^2) = log_v(v_i).
+//
+// The Fiat-Shamir scheme works as follows in our case:
+//
+// ZKP construction
+//
+// - Pick random r
+// - Compute E as:
+//   E = HASH(a, b, c^4, c_i^2 ), where
+//     a = (c^4)^r mod n^2
+//     b = V^r mod n^2
+//     c is a cyphertext,
+//     V is a generator from ThresholdKey.V
+//     c_i is a partial decryption for this server
+// - Compute Z as:
+//    delta * E * s_i + r, where
+//      delta is the factorial of the number of decryption servers
+//      s_i is a secret share for this server
+//
+// ZKP verification
+//
+// - Compute the original a from
+//   a = a1 * a2 mod n^2
+//   a1 = (c^4)^Z
+//   a2 = [ (c_i^2)^E ]^-1 mod n^2
+// - Compute the original b from
+//   b = b1 * b2 mod n^2
+//   b1 = V^Z
+//   b2 = [ v_i^E ] -1 mod n^2
+// - Rehash H(a, b, c^4, c_i)
+// - Compare ZKP hash with the one just computed
```

Comment:
One thing I don't entirely follow here: the verification hash uses `c_i` vs the construction hash uses `c_i^2`. How come? Also, is `c_i` different for each the verifying participant vs the one who constructs the proof?

Reply:
You are right, it's a typo. We should have `c_i^2` there.

> Also, is c_i different for each the verifying participant vs the one who constructs the proof?

That's correct. What we do here is proving that me, decryption server `i` indeed used mine secret exponent `s_i` for partial decryption without revealing `s_i`. We basically want to eliminate servers wanting to mess up. 

The server combining shares has its own `s_i` and it's different from other servers. However, we have the partial `c_i` decryption stored in `PartialDecryption` struct as `Decryption` and that's the value we use to verify ZKP (see the second line of `verifyPart1`).